wagtail/contrib/settings/models.py

```diff
@@ -161,51 +219,84 @@ class Meta:
-        return first_obj
+    def _get_or_create(cls, **kwargs):
+        raise NotImplementedError
 
     @classmethod
-    def load(cls, request_or_site=None):
+    def load(cls, request_or_site=None, locale=None):
         """
-        Get or create an instance of this model. There is only ever one
-        instance of models inheriting from `AbstractSetting` so we can
-        use `pk=1`.
-
+        Get or create an instance of this model.
         If `request_or_site` is present and is a request object, then we cache
         the result on the request for faster repeat access.
         """
-
         # We can only cache on the request, so if there is no request then
         # we know there's nothing in the cache.
         if request_or_site is None or isinstance(request_or_site, Site):
-            return cls._get_or_create()
+            return cls._get_or_create(locale=locale)
 
         # Check if we already have this in the cache and return it if so.
         attr_name = cls.get_cache_attr_name()
         if hasattr(request_or_site, attr_name):
             return getattr(request_or_site, attr_name)
 
-        obj = cls._get_or_create()
+        locale = cls._get_locale(locale=locale, request=request_or_site)
+        obj = cls._get_or_create(locale=locale)
 
         # Cache for next time.
         setattr(request_or_site, attr_name, obj)
-
         return obj
 
     def __str__(self):
         return self._meta.verbose_name
 
 
+class BaseGenericSetting(AbstractGenericSetting):
+    class Meta:
+        abstract = True
+
+    @classmethod
+    def _get_or_create(cls, **kwargs):
+        """
+        Internal convenience method to get or create the first instance.
+
+        We cannot hardcode `pk=1`, for example, as not all database backends
+        use sequential IDs (e.g. Postgres).
+        """
+
+        first_obj = cls.base_queryset().first()
+        if first_obj is None:
+            return cls.objects.create()
+        return first_obj
+
+
+class BaseTranslatableGenericSetting(TranslatableMixin, AbstractGenericSetting):
+    # The translation key should be derived from a constant
+    # because we want only one instance per locale,
+    # so the first instance's ID (=1) is basically the translation key.
+    _translation_key = uuid.uuid5(
+        uuid.UUID("4e47faf7-d91f-411f-8a8f-51a05d75f992"), "1"
+    )
+
+    class Meta:
+        abstract = True
+        unique_together = [("translation_key", "locale")]
+
+    @classmethod
+    def _get_locale(cls, locale, request):
+        return locale or get_locale_for(request=request, model=cls)
+
+    @classmethod
+    def _get_or_create(cls, locale=None):
+        """
+        Internal convenience method to get or create the instance associated
+        to the locale requested.
+        """
+        return cls.base_queryset().get_or_create(
+            locale=locale or Locale.get_active(),
+            translation_key=cls._translation_key,
```

Comment:
This will not find the alternative version if target items have `translation_key` values that don't fit our 'fixed formula' - which I don't think we really have any grounds to impose... the migration path for existing settings models will likely be very similar to that of Snippet models, where `translation_key` values are first set in a data migration, and are just randomly generated at that point. We need to support that.

`TranslatableMixin` has an existing solution for getting the appropriate translation in this case: the `localized` property, and it doesn't require access to the request or a separately provided `Locale` object. We should find a way to utlise this here and try to keep the rest of the API simple..

Reply:
FYI, here's what worked in the backported version - a simple override to `get_site()`:

```
    @classmethod
    def for_site(cls, site, locale=None):
        """
        Overrides BaseSiteSetting.for_site() to get or create settings for the
        'provided' locale when editing OR the 'active' locale when retrieving
        an instance to display to users.
        """
        instance, created = cls.base_queryset().get_or_create(
            site=site,
            locale=locale or Locale.get_active(),
            defaults={"translation_key": get_default_translation_key(site)},
        )
        return instance
```

Reply:
On second thoughts. I don't think we really do want to automatically create versions in the active language when requesting from the front-end, so I have evolved the above to the following in a client project:

```
@classmethod
def for_site(cls, site, locale=None):
    ideal_translation_key = get_default_translation_key(site)

    if locale:
        # For requests from the admin, we do always want to
        # get or create an instance in the supplied locale
        instance, _ = cls.base_queryset().get_or_create(
            site=site,
            locale=locale,
            defaults={"translation_key": ideal_tranlsation_key},
        )

    # For requests from the front-end, we do not want to automatically
    # create objects in the selected language when they do not yet exist
    # In those cases: We want to return the default language version
    # instead (or create that if it doesn't exist)

    active_locale = Locale.get_active()
    default_locale = Locale.get_default()
    default_language_version = None

    for instance in cls.base_queryset().filter(
        site=site,
        locale__in=[
            active_locale,
            default_locale,
        ],
    ):
        if instance.locale_id == active_locale.id:
            # Hoorah! we have our preferred return value
            return instance
        else:
            default_language_version = instance

    # Return the default language version if it exists
    if default_language_version:
        return default_language_version

    # Or create it if it doesn't
    return cls.base_queryset().create(
        site=site,
        locale=default_locale,
        translation_key=ideal_translation_key,
    )

```

Reply:
On reflection, a lot of this complication comes from the fact the view is using this method, when it doesn't really have to. We can dedicate this method to finding an object for the front-end to use, and shift the rest into the Admin view.